### PR TITLE
[Feat] Add MVMoE model

### PR DESCRIPTION
**Pull request by RoyalSkye**
## Description

This PR adds support for [`MVMoE`](https://arxiv.org/pdf/2405.01029) in the multi-task VRP setting. In specific, the token-level [input-choice](https://arxiv.org/pdf/1701.06538) gating is implemented. Some minor bugs are fixed as well, such as missing the definition of `MTVRPInitEmbedding` and `MTVRPContext`. 

* [x] An example will be given in `1-mtvrp.ipynb`.

## Motivation and Context

A step towards foundation VRP models.

- [ ] I have raised an issue to propose this change ([required](https://github.com/ai4co/rl4co/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: rl4co/models/nn/graph/attnnet.py [hunk 1]
```diff
@@ -29,19 +32,18 @@ def __init__(
         feedforward_hidden: int = 512,
         normalization: Optional[str] = "batch",
         sdpa_fn: Optional[Callable] = None,
+        moe_kwargs: Optional[dict] = None,
```
Comment (cbhua) [inline rl4co/models/nn/graph/attnnet.py:35]: Same as the pointer attention, we may want a separate class for MoE's MHA.

File: rl4co/models/nn/env_embeddings/init.py [hunk 2]
```diff
@@ -444,3 +445,25 @@ def _stepwise_operations_embed(self, td: TensorDict):
...
+        # [!] convert [0, inf] -> [0, 0] if a problem does not include the time window constraint, do not modify in-place
+        time_windows = torch.where(torch.isinf(time_windows), torch.tensor(0.), time_windows)
+        # embeddings
+        depot_embedding = self.init_embed_depot(depot)
+        node_embeddings = self.init_embed(
+            torch.cat(
+                (cities, demand_linehaul[..., None], demand_backhaul[..., None], time_windows, service_time[..., None]), -1
+            )
+        )
+        return torch.cat((depot_embedding, node_embeddings), -2)
```
Comment (cbhua) [inline rl4co/models/nn/env_embeddings/init.py:469]: Do we need to record the moe loss here?

**Review by cbhua — COMMENTED** [review]
Great job! Thanks for your work! Left some discussion points~ 😁

File: rl4co/models/nn/env_embeddings/init.py (hunk 2 above)
Comment (fedebotu) [reply rl4co/models/nn/env_embeddings/init.py:469]: > Do we need to record the moe loss here?

If we use the raw feature MoE yes, but @RoyalSkye showed this does not work too well, so we might skip I guess?

File: rl4co/models/nn/env_embeddings/init.py [hunk 3]
```diff
@@ -444,3 +445,25 @@ def _stepwise_operations_embed(self, td: TensorDict):
...
+        # node_dim = 7: x, y, demand_linehaul, demand_backhaul, tw start, tw end, service time
+        super(MTVRPInitEmbedding, self).__init__(embed_dim, linear_bias, node_dim)
+
+    def forward(self, td):
+        depot, cities = td["locs"][:, :1, :], td["locs"][:, 1:, :]
+        demand_linehaul, demand_backhaul = td["demand_linehaul"][..., 1:], td["demand_backhaul"][..., 1:]
+        service_time = td["service_time"][..., 1:]
+        time_windows = td["time_windows"][..., 1:, :]
+        # [!] convert [0, inf] -> [0, 0] if a problem does not include the time window constraint, do not modify in-place
+        time_windows = torch.where(torch.isinf(time_windows), torch.tensor(0.), time_windows)
```
Comment (fedebotu) [inline rl4co/models/nn/env_embeddings/init.py:461]: The same can be achieved slightly more efficiently with:

```python
 time_windows = torch.nan_to_num(time_windows,  posinf=0.0)
```

File: rl4co/models/zoo/mvmoe/model.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                "normalization": "instance",
+                "use_graph_context": False,
+                "moe_kwargs": moe_kwargs,
+            }
+            policy_kwargs.update(policy_kwargs_)
+            policy = AttentionModelPolicy(env_name=env.name, **policy_kwargs)
+
+        # Initialize with the shared baseline
+        super(MVMoE_POMO, self).__init__(env, policy, policy_kwargs, baseline, num_augment, augment_fn,
+                                         first_aug_identity, feats, num_starts, **kwargs)
```
Comment (fedebotu) [inline rl4co/models/zoo/mvmoe/model.py:49]: It seems that the MoE loss is not included, right?
(I sent you on Slack a reference)

File: rl4co/models/nn/attention.py [hunk 5]
```diff
@@ -145,6 +146,7 @@ class PointerAttention(nn.Module):
         linear_bias: whether to use bias in linear projection
         check_nan: whether to check for NaNs in logits
         sdpa_fn: scaled dot product attention function (SDPA) implementation
+        moe_kwargs: Keyword arguments for MoE
     """
 
     def __init__(
```
[the remark this answers is not in the document]
Comment (fedebotu) [reply rl4co/models/nn/attention.py:152]: Mmmh... In principle, it might be better to have all the contributions of a model in its own repo. But given that such contribution could improve other models (say: if you wanted, you may try applying MoE to scheduling) and **if it does not change the base performance** (which seems to be the case here?), then I think it can be good. What do you think?

Asking also master @LTluttmann

File: rl4co/models/zoo/mvmoe/model.py (hunk 4 above)
Comment (RoyalSkye) [reply rl4co/models/zoo/mvmoe/model.py:49]: Yes, we recently found the MoE loss may conflict with the reinforcement loss during optimization, and thus I ignored it. Should we add it?

File: rl4co/models/nn/env_embeddings/init.py (hunk 2 above)
Comment (RoyalSkye) [reply rl4co/models/nn/env_embeddings/init.py:469]: Please see the below comments.

File: rl4co/models/nn/attention.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (RoyalSkye) [reply rl4co/models/nn/attention.py:152]: Yes, maybe I can create a new `class PointerAttention_MoE(PointerAttention)`. Let me know your final decision.

File: rl4co/models/zoo/mvmoe/model.py (hunk 4 above)
Comment (cbhua) [reply rl4co/models/zoo/mvmoe/model.py:49]: If the MoE loss doesn't contribute much to the performance I think we could just skip it.

File: rl4co/models/nn/attention.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (LTluttmann) [reply rl4co/models/nn/attention.py:152]: I would also go for a separate `PointerAttnMoE(PointerAttention)` class for MoE, which would only need to overwrite the `project_out`. I guess we have the best of both worlds then - most of the `PointerAttention` code is reused while we separate the logic in a clear manner.

[the remark this answers is not in the document]
Comment (RoyalSkye) [reply rl4co/models/nn/attention.py:152]: okay, I will finish this part in the next commit!

[the remark this answers is not in the document]
Comment (RoyalSkye) [reply rl4co/models/nn/attention.py]: Hi all, I have finished this part. As for the MHA in `attnnet.py`, I think the current code is neat, so I keep it as is. In my opinion, MoE is a general idea, which may benefit various CO tasks. So implementing it in the underlying code may be a better choice..? What do you think? Kindly let me know if any problem exists.

[the remark this answers is not in the document]
Comment (fedebotu) [reply rl4co/models/nn/attention.py]: On that I agree, it doesn't impact performance in MHA in `attnnet.py` correct? If so, I'm good keeping it, so people may explore it!

[the remark this answers is not in the document]
Comment (RoyalSkye) [reply rl4co/models/nn/attention.py]: Yes, the default setting is `moe_kwargs=None`. In this case, it's equivalent to the original MHA, and thus the performance will not be affected.

File: rl4co/models/zoo/am/decoder.py [hunk 6]
```diff
@@ -108,7 +108,7 @@ def __init__(
 
         if pointer is None:
             # MHA with Pointer mechanism (https://arxiv.org/abs/1506.03134)
-            pointer = PointerAttention(
```
Comment (fedebotu) [inline rl4co/models/zoo/am/decoder.py:109]: So we are defaulting to `PointerAttnMoE` here for the base AM class? I think this might be confusing for some who just want the default model (without MoE). I would rather keep the previous version in this case (`PointerAttention` but with MoE kwargs)

Or, I would suggest to do this:
```python
pointer_attn_class = PointerAttention if moe_kwargs is None else PointerAttnMoE
pointer = pointer_attn_class(embed_dim ...)
```

File: rl4co/models/nn/attention.py [hunk 7]
```diff
@@ -220,6 +211,53 @@ def _inner_mha(self, query, key, value, attn_mask):
...
+        if self.moe_kwargs is not None:
+            self.project_out = None
+            self.project_out_moe = MoE(embed_dim, embed_dim, num_neurons=[], out_bias=out_bias, **moe_kwargs)
+            if self.moe_kwargs["light_version"]:
+                self.dense_or_moe = nn.Linear(embed_dim, 2, bias=False)
+                self.project_out = nn.Linear(embed_dim, embed_dim, bias=out_bias)
+
     def _project_out(self, out):
         """Implementation of Hierarchical Gating based on Zhou et al. (2024) <https://arxiv.org/abs/2405.01029>."""
         if self.moe_kwargs is not None:
```
Comment (fedebotu) [inline rl4co/models/nn/attention.py:263]: (read comment below) if `PointerAttnMoe` by default has `MoE`, then we could also remove this "if, else" statement

File: rl4co/models/zoo/am/decoder.py (hunk 6 above)
Comment (RoyalSkye) [reply rl4co/models/zoo/am/decoder.py:109]: The second one looks good! Please see the latest commit.

**Review by fedebotu — APPROVED** [review]
:rocket: 

@cbhua @LTluttmann any of you who is the last one to approve can merge!

**Review by cbhua — APPROVED** [review]
Nice work, let go~

**Comment by RoyalSkye** [comment]
I have added an example to `1-mtvrp.ipynb`. It seems that the notebook has some issues on my laptop, so I just uploaded a raw version without output 🚀.

**Comment by fedebotu** [comment]
> It seems that the notebook has some issues on my laptop, so I just uploaded a raw version without output 

Are you using an M-series Macbook? We had some problems only with MTVRP and only with Apple MPS accelerators :rofl:

**Comment by RoyalSkye** [comment]
Thank Chuanbo. Looks great! I have nothing to add to this PR.

**Comment by cbhua** [comment]
Great job! After @LTluttmann having a look let's launch! 🚀

**Comment by LTluttmann** [comment]
awesome work! Clean and well documented code, was a pleasure to read it haha! Merging now
